src/include/stat.h

```diff
@@ -6,35 +6,169 @@
  * See the file LICENSE for redistribution information.
  */
 
+/*
+ * ARRAY COUNTERS: DESIGN DESCRIPTION.
+ *
+ * Instead of a single counter we use an array of values.
+ * Threads update different values in an array to avoid writing the same
+ * cache line and incurring the cache coherency overheads, which can
+ * dramatically slowdown fast and otherwise read-mostly
+ * workloads. Upon reading a value, items in individual array cells
+ * are merged and returned to the caller.
+ * Aggregation is performed without locking, so the counter read may be
+ * slightly inconsistent. Existing use cases tolerate eventual consistency.
+ * We are not making things worse relative to non-array implementation.
+ * The read operation now takes longer, because of aggregation, so we
+ * are slightly increasing the window when things may be inconsistent,
+ * but that's about it.
+ *
+ * We used a fixed number of slots in an array. Picking the number of slots
+ * is not as straightforward as it should be.
+ * Ideally, if our application is running on the system alone is CPU-intensive,
+ * and is using all CPUs on the system, we want to use the same number of slots
+ * as there are CPUs (because their L1 caches are the units of coherency).
+ * However, in practice we cannot easily determine how many CPUs are actually
+ * available for the application.
+ * Our next best option is to use the number of threads in the application as
+ * a heuristic for the number of CPUs. Unfortunately, However, inside WT we do
+ * not know when the application creates its threads.
+ *
+ * Our solution is to simply use a fixed number of slots that roughly
+ * approximates the largest number of cores we expect to see on the machine
+ * where WT is used. All we want is to avoid all threads writing the same
+ * variable at the same time, and that solution does the trick.
+ *
+ */
+#define	WT_COUNTER_SLOTS 24
```

Comment:
@fedorova, I'll update it -- once I think we're ready to pull, I'll ask you to do one more pass to make sure I've not broken anything.